src/main/java/com/yelp/nrtsearch/server/luceneserver/RegisterFieldsHandler.java

```diff
@@ -111,16 +121,28 @@ public FieldDefResponse handle(final IndexState indexState, FieldDefRequest fiel
           throw new RuntimeException(e);
         }
 
-        pendingFieldDefs.put(
-            fieldName, parseOneFieldType(indexState, pendingFieldDefs, fieldName, currentField));
+        FieldDef fieldDef =
+            parseOneFieldType(indexState, pendingFieldDefs, fieldName, currentField);
+        if (currentField.getKey() && !(fieldDef instanceof Keyable)) {
+          throw new RegisterFieldsException(
+              "field \"" + fieldName + "\" is not a keyable field type");
+        }
+        pendingFieldDefs.put(fieldName, fieldDef);
       }
     }
 
     // add FieldDef and its corresponding JsonObject to states variable in IndexState
     for (Map.Entry<String, FieldDef> ent : pendingFieldDefs.entrySet()) {
       JsonObject fieldAsJsonObject =
           jsonParser.parse(saveStates.get(ent.getKey())).getAsJsonObject();
-      indexState.addField(ent.getValue(), fieldAsJsonObject);
+      FieldDef fieldDef = ent.getValue();
+      indexState.addField(fieldDef, fieldAsJsonObject);
+      if (fieldDef instanceof Keyable) {
+        Keyable keyableField = (Keyable) fieldDef;
+        if (keyableField.isKey()) {
+          indexState.setKeyableField(keyableField);
+        }
+      }
```

Comment:
This could also be done by holding a reference to the key FieldDef when building it above, since there will be at most one field.

Reply:
Makes sense, I have changed the code to reflect this.